# 01 - Data Understanding

Initial data inspection, schema validation, summary statistics, and missing value analysis.

In [2]:
import pandas as pd

file_path = "../data/raw/Fraud.csv"

df = pd.read_csv(file_path)

print(df.shape)
print(df.head())
print(df.info())
print(df["isFraud"].value_counts())
print(df["isFraud"].value_counts(normalize=True))

(6362620, 11)
   step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data column

In [3]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", lambda value: f"{value:,.4f}")

In [4]:
DATA_PATH = "../data/raw/Fraud.csv"

df = pd.read_csv(DATA_PATH)

print(f"Dataset loaded successfully.")
print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")

Dataset loaded successfully.
Rows: 6,362,620
Columns: 11


In [7]:
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


In [8]:
schema = pd.DataFrame({
    "column": df.columns,
    "dtype": df.dtypes.astype(str).values,
    "non_null_count": df.notna().sum().values,
    "missing_count": df.isna().sum().values,
    "unique_values": df.nunique().values
})

schema

,column,dtype,non_null_count,missing_count,unique_values
0,step,int64,6362620,0,743
1,type,object,6362620,0,5
2,amount,float64,6362620,0,5316900
3,nameOrig,object,6362620,0,6353307
4,oldbalanceOrg,float64,6362620,0,1845844
5,newbalanceOrig,float64,6362620,0,2682586
6,nameDest,object,6362620,0,2722362
7,oldbalanceDest,float64,6362620,0,3614697
8,newbalanceDest,float64,6362620,0,3555499
9,isFraud,int64,6362620,0,2


In [9]:
missing_values = (
    df.isna()
      .sum()
      .rename("missing_count")
      .to_frame()
)

missing_values["missing_percentage"] = (
    missing_values["missing_count"] / len(df) * 100
)

missing_values

,missing_count,missing_percentage
step,0,0.0000
type,0,0.0000
amount,0,0.0000
nameOrig,0,0.0000
oldbalanceOrg,0,0.0000
newbalanceOrig,0,0.0000
nameDest,0,0.0000
oldbalanceDest,0,0.0000
newbalanceDest,0,0.0000
isFraud,0,0.0000


In [10]:
total_missing = df.isna().sum().sum()

print(f"Total missing values: {total_missing:,}")

Total missing values: 0


In [11]:
duplicate_count = df.duplicated().sum()
duplicate_percentage = duplicate_count / len(df) * 100

print(f"Duplicate rows: {duplicate_count:,}")
print(f"Duplicate percentage: {duplicate_percentage:.6f}%")

Duplicate rows: 0
Duplicate percentage: 0.000000%


In [12]:
numerical_summary = df.describe().T

numerical_summary

,count,mean,std,min,25%,50%,75%,max
step,"6,362,620.0000",243.3972,142.3320,1.0000,156.0000,239.0000,335.0000,743.0000
amount,"6,362,620.0000","179,861.9035","603,858.2315",0.0000,"13,389.5700","74,871.9400","208,721.4775","92,445,516.6400"
oldbalanceOrg,"6,362,620.0000","833,883.1041","2,888,242.6730",0.0000,0.0000,"14,208.0000","107,315.1750","59,585,040.3700"
newbalanceOrig,"6,362,620.0000","855,113.6686","2,924,048.5030",0.0000,0.0000,0.0000,"144,258.4100","49,585,040.3700"
oldbalanceDest,"6,362,620.0000","1,100,701.6665","3,399,180.1130",0.0000,0.0000,"132,705.6650","943,036.7075","356,015,889.3500"
newbalanceDest,"6,362,620.0000","1,224,996.3982","3,674,128.9421",0.0000,0.0000,"214,661.4400","1,111,909.2500","356,179,278.9200"
isFraud,"6,362,620.0000",0.0013,0.0359,0.0000,0.0000,0.0000,0.0000,1.0000
isFlaggedFraud,"6,362,620.0000",0.0000,0.0016,0.0000,0.0000,0.0000,0.0000,1.0000


In [13]:
total_transactions = len(df)
fraud_transactions = int(df["isFraud"].sum())
legitimate_transactions = total_transactions - fraud_transactions

fraud_rate = fraud_transactions / total_transactions
legitimate_rate = legitimate_transactions / total_transactions

print(f"Total transactions:      {total_transactions:,}")
print(f"Fraudulent transactions: {fraud_transactions:,}")
print(f"Legitimate transactions: {legitimate_transactions:,}")
print(f"Fraud rate:              {fraud_rate:.6%}")
print(f"Legitimate rate:         {legitimate_rate:.6%}")

Total transactions:      6,362,620
Fraudulent transactions: 8,213
Legitimate transactions: 6,354,407
Fraud rate:              0.129082%
Legitimate rate:         99.870918%


In [14]:
total_transaction_value = df["amount"].sum()

fraudulent_transaction_value = df.loc[
    df["isFraud"] == 1,
    "amount"
].sum()

legitimate_transaction_value = df.loc[
    df["isFraud"] == 0,
    "amount"
].sum()

average_fraud_amount = df.loc[
    df["isFraud"] == 1,
    "amount"
].mean()

average_legitimate_amount = df.loc[
    df["isFraud"] == 0,
    "amount"
].mean()

fraud_value_percentage = (
    fraudulent_transaction_value / total_transaction_value
)

print(f"Total transaction value:       {total_transaction_value:,.2f}")
print(f"Fraudulent transaction value:  {fraudulent_transaction_value:,.2f}")
print(f"Legitimate transaction value:  {legitimate_transaction_value:,.2f}")
print(f"Average fraudulent amount:     {average_fraud_amount:,.2f}")
print(f"Average legitimate amount:     {average_legitimate_amount:,.2f}")
print(f"Fraudulent share of value:     {fraud_value_percentage:.6%}")

Total transaction value:       1,144,392,944,759.77
Fraudulent transaction value:  12,056,415,427.84
Legitimate transaction value:  1,132,336,529,331.93
Average fraudulent amount:     1,467,967.30
Average legitimate amount:     178,197.04
Fraudulent share of value:     1.053521%


In [15]:
median_fraud_amount = df.loc[
    df["isFraud"] == 1,
    "amount"
].median()

median_legitimate_amount = df.loc[
    df["isFraud"] == 0,
    "amount"
].median()

print(f"Median fraudulent amount: {median_fraud_amount:,.2f}")
print(f"Median legitimate amount: {median_legitimate_amount:,.2f}")

Median fraudulent amount: 441,423.44
Median legitimate amount: 74,684.72


In [16]:
minimum_step = df["step"].min()
maximum_step = df["step"].max()
number_of_steps = df["step"].nunique()

print(f"Minimum step:    {minimum_step}")
print(f"Maximum step:    {maximum_step}")
print(f"Unique steps:    {number_of_steps}")
print(f"Approximate days represented: {number_of_steps / 24:.2f}")

Minimum step:    1
Maximum step:    743
Unique steps:    743
Approximate days represented: 30.96


In [17]:
transactions_by_type = (
    df.groupby("type", observed=True)
      .agg(
          transaction_count=("type", "size"),
          total_value=("amount", "sum"),
          average_amount=("amount", "mean"),
          median_amount=("amount", "median")
      )
      .sort_values("transaction_count", ascending=False)
)

transactions_by_type["transaction_percentage"] = (
    transactions_by_type["transaction_count"]
    / total_transactions
    * 100
)

transactions_by_type

,transaction_count,total_value,average_amount,median_amount,transaction_percentage
type,,,,,
CASH_OUT,2237500,"394,412,995,224.4900","176,273.9643","147,072.1850",35.1663
PAYMENT,2151495,"28,093,371,138.3700","13,057.6047","9,482.1900",33.8146
CASH_IN,1399284,"236,367,391,912.4600","168,920.2420","143,427.7100",21.9923
TRANSFER,532909,"485,291,987,263.1700","910,647.0096","486,308.3900",8.3756
DEBIT,41432,"227,199,221.2800","5,483.6653","3,048.9900",0.6512


In [19]:
df_analysis = df.assign(
    fraudulent_amount=df["amount"] * df["isFraud"]
)

fraud_summary_by_type = (
    df_analysis.groupby("type", observed=True)
      .agg(
          transaction_count=("type", "size"),
          fraud_count=("isFraud", "sum"),
          total_value=("amount", "sum"),
          fraudulent_value=("fraudulent_amount", "sum")
      )
)

fraud_summary_by_type["fraud_rate"] = (
    fraud_summary_by_type["fraud_count"]
    / fraud_summary_by_type["transaction_count"]
)

fraud_summary_by_type["fraud_share_of_value"] = (
    fraud_summary_by_type["fraudulent_value"]
    / fraud_summary_by_type["total_value"]
)

fraud_summary_by_type = fraud_summary_by_type.sort_values(
    "fraud_count",
    ascending=False
)

fraud_summary_by_type

,transaction_count,fraud_count,total_value,fraudulent_value,fraud_rate,fraud_share_of_value
type,,,,,,
CASH_OUT,2237500,4116,"394,412,995,224.4900","5,989,202,243.8300",0.0018,0.0152
TRANSFER,532909,4097,"485,291,987,263.1700","6,067,213,184.0100",0.0077,0.0125
CASH_IN,1399284,0,"236,367,391,912.4600",0.0000,0.0000,0.0000
DEBIT,41432,0,"227,199,221.2800",0.0000,0.0000,0.0000
PAYMENT,2151495,0,"28,093,371,138.3700",0.0000,0.0000,0.0000


In [20]:
fraud_summary_display = fraud_summary_by_type.copy()

fraud_summary_display["fraud_rate"] = (
    fraud_summary_display["fraud_rate"] * 100
)

fraud_summary_display["fraud_share_of_value"] = (
    fraud_summary_display["fraud_share_of_value"] * 100
)

fraud_summary_display = fraud_summary_display.rename(columns={
    "transaction_count": "Transactions",
    "fraud_count": "Frauds",
    "total_value": "Total Value",
    "fraudulent_value": "Fraudulent Value",
    "fraud_rate": "Fraud Rate (%)",
    "fraud_share_of_value": "Fraud Share of Value (%)"
})

fraud_summary_display

,Transactions,Frauds,Total Value,Fraudulent Value,Fraud Rate (%),Fraud Share of Value (%)
type,,,,,,
CASH_OUT,2237500,4116,"394,412,995,224.4900","5,989,202,243.8300",0.1840,1.5185
TRANSFER,532909,4097,"485,291,987,263.1700","6,067,213,184.0100",0.7688,1.2502
CASH_IN,1399284,0,"236,367,391,912.4600",0.0000,0.0000,0.0000
DEBIT,41432,0,"227,199,221.2800",0.0000,0.0000,0.0000
PAYMENT,2151495,0,"28,093,371,138.3700",0.0000,0.0000,0.0000


In [21]:
actual_frauds = df["isFraud"] == 1
flagged_transactions = df["isFlaggedFraud"] == 1

correctly_flagged_frauds = (actual_frauds & flagged_transactions).sum()
missed_frauds = (actual_frauds & ~flagged_transactions).sum()
incorrectly_flagged_legitimate = (~actual_frauds & flagged_transactions).sum()

existing_rule_recall = (
    correctly_flagged_frauds / fraud_transactions
)

flagged_count = flagged_transactions.sum()

existing_rule_precision = (
    correctly_flagged_frauds / flagged_count
    if flagged_count > 0
    else 0
)

print(f"Transactions flagged:            {flagged_count:,}")
print(f"Correctly flagged frauds:         {correctly_flagged_frauds:,}")
print(f"Missed frauds:                    {missed_frauds:,}")
print(f"Incorrectly flagged legitimate:   {incorrectly_flagged_legitimate:,}")
print(f"Existing rule recall:             {existing_rule_recall:.4%}")
print(f"Existing rule precision:          {existing_rule_precision:.4%}")

Transactions flagged:            16
Correctly flagged frauds:         16
Missed frauds:                    8,197
Incorrectly flagged legitimate:   0
Existing rule recall:             0.1948%
Existing rule precision:          100.0000%


In [22]:
flagged_fraud_value = df.loc[
    actual_frauds & flagged_transactions,
    "amount"
].sum()

missed_fraud_value = df.loc[
    actual_frauds & ~flagged_transactions,
    "amount"
].sum()

fraud_value_recall = (
    flagged_fraud_value / fraudulent_transaction_value
)

print(f"Fraudulent value flagged: {flagged_fraud_value:,.2f}")
print(f"Fraudulent value missed:  {missed_fraud_value:,.2f}")
print(f"Fraud value recall:       {fraud_value_recall:.4%}")

Fraudulent value flagged: 77,785,563.69
Fraudulent value missed:  11,978,629,864.15
Fraud value recall:       0.6452%


In [23]:
headline_metrics = pd.DataFrame({
    "metric": [
        "total_transactions",
        "fraudulent_transactions",
        "legitimate_transactions",
        "fraud_rate",
        "total_transaction_value",
        "fraudulent_transaction_value",
        "fraudulent_share_of_value",
        "average_fraud_amount",
        "average_legitimate_amount",
        "median_fraud_amount",
        "median_legitimate_amount",
        "existing_rule_recall",
        "existing_rule_precision",
        "existing_rule_value_recall"
    ],
    "value": [
        total_transactions,
        fraud_transactions,
        legitimate_transactions,
        fraud_rate,
        total_transaction_value,
        fraudulent_transaction_value,
        fraud_value_percentage,
        average_fraud_amount,
        average_legitimate_amount,
        median_fraud_amount,
        median_legitimate_amount,
        existing_rule_recall,
        existing_rule_precision,
        fraud_value_recall
    ]
})

headline_metrics

,metric,value
0,total_transactions,"6,362,620.0000"
1,fraudulent_transactions,"8,213.0000"
2,legitimate_transactions,"6,354,407.0000"
3,fraud_rate,0.0013
4,total_transaction_value,"1,144,392,944,759.7700"
5,fraudulent_transaction_value,"12,056,415,427.8400"
6,fraudulent_share_of_value,0.0105
7,average_fraud_amount,"1,467,967.2991"
8,average_legitimate_amount,"178,197.0417"
9,median_fraud_amount,"441,423.4400"


In [ ]:
OUTPUT_PATH = "../data/processed/headline_metrics.csv"

headline_metrics.to_csv(OUTPUT_PATH, index=False)

## Initial Findings

1. Fraud represents only 0.1291% of transactions but approximately
   1.05% of transaction value, demonstrating that fraud is rare but
   financially disproportionate.

2. The average fraudulent transaction is approximately 8.2 times
   larger than the average legitimate transaction. However, amount
   alone cannot establish whether a transaction is fraudulent.

3. Fraud occurs only in TRANSFER and CASH_OUT transactions within
   PaySim. TRANSFER has approximately 4.2 times the fraud rate of
   CASH_OUT, while the two types contribute nearly equal fraudulent value.

4. The existing rule achieves 100% precision but only approximately
   0.19% recall, indicating that it produces highly reliable alerts but
   misses almost all fraudulent transactions.

5. The dataset contains no missing values or exact duplicate rows.
   This establishes structural completeness but does not guarantee
   transactional validity or real-world representativeness.